Libplanet/Blockchain/BlockChain.cs

```diff
@@ -1095,26 +1097,18 @@ public BlockChain<T> Fork(BlockHash point, bool inheritRenderers = true)
         /// <returns>A instance of block locator.</returns>
         public BlockLocator GetBlockLocator(int threshold = 10)
         {
-            Guid chainId;
-            BlockHash tipHash;
-
             _rwlock.EnterReadLock();
             try
             {
-                chainId = Id;
-                tipHash = Tip.Hash;
+                return BlockLocator.Create(
```

Comment:
Well...  I'm still convinced that lock separation on here wasn't a trade-off between safety and performance, but was a sort of correction against unnecessary locking. (because we keep the immutability on previous blocks for a given chain id).

Reply:
> because we keep the immutability on previous blocks for a given chain id

This is only implicitly guaranteed by our current practice/codebase. I do agree there is little to no danger in how we are currently using `IStore` and `BlockChain<T>`, but the `IStore` interface itself does not contractually guarantee such immutability (if we had something like `IWriteOnlyStore` with appropriate constraints, then I'd say we *really* don't need to worry about safety). The immutability assumption only comes from *knowing* the whole stack all the way up to `Swarm<T>` is not going to mess with "bad" `IStore` methods.

There is also the question of how we specify `BlockLocator`'s behavior. One of the following should apply:
- If `IStore` is assumed to be immutable on all previous `Block<T>`s for the current `BlockChain<T>.Tip` *and* `BlockChain<T>` (and in turn the `IStore`) is assumed to have *all* `Block<T>`s, then `BlockLocator` should not handle `null` for missing `Block<T>`s.
- If `IStore` is not assumed to be immutable or could be missing some `Block<T>`s, then `BlockLocator` should be designed to deal with missing `Block<T>`s.

The problem is, we have the latter as an implementation while assuming the former via usage (of `BlockChain<T>` and `IStore` *not* `BlockLocator`, which can be confusing). 🙄

Reply:
- In the long term, I hope `BlockLocator` disappears and `BlockChain<T>` and `IStore` adjust together. but I don't think this is something that can be done right now.
- If the lock range can be reduced by tweaking `IStore`'s interface, I think it's worth considering for at least now.  
  - I'm not a fan of the Chain ID concept and think a universal storage interface is needed
  - but also I think it's expensive to pay for reducible locks until that happens.